File: airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/book_keeping/events/AirbyteStreamStatusMessageEventListener.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.internal.book_keeping.events;

import io.airbyte.protocol.models.AirbyteMessage;
import io.airbyte.protocol.models.AirbyteMessage.Type;
import io.airbyte.protocol.models.AirbyteTraceMessage;
import io.airbyte.workers.internal.book_keeping.StreamStatusTracker;
import io.micronaut.context.event.ApplicationEventListener;
import io.micronaut.scheduling.annotation.Async;
import jakarta.inject.Singleton;

/**
 * Custom application listener that handles Airbyte Protocol {@link Type#TRACE} messages of type
 * {@link AirbyteTraceMessage.Type#STREAM_STATUS} to track the status (or progress) of a stream
 * within replication. The listener handles the messages asynchronously to avoid blocking
 * replication.
 */
@Singleton
public class AirbyteStreamStatusMessageEventListener implements ApplicationEventListener<ReplicationAirbyteMessageEvent> {

  private final StreamStatusTracker streamStatusTracker;

  public AirbyteStreamStatusMessageEventListener(final StreamStatusTracker streamStatusTracker) {
    this.streamStatusTracker = streamStatusTracker;
  }

  @Override
  @Async("stream-status")
  public void onApplicationEvent(final ReplicationAirbyteMessageEvent event) {
    streamStatusTracker.track(event);
  }

  @Override
  public boolean supports(final ReplicationAirbyteMessageEvent event) {
    return isStreamStatusTraceMessage(event.airbyteMessage());
  }

  private boolean isStreamStatusTraceMessage(final AirbyteMessage airbyteMessage) {
    return Type.TRACE.equals(airbyteMessage.getType())
        && AirbyteTraceMessage.Type.STREAM_STATUS.equals(airbyteMessage.getTrace().getType());
  }

}

File: airbyte-commons-worker/src/main/kotlin/io/airbyte/workers/internal/bookkeeping/events/AirbyteStreamStatusMessageEventListener.kt
package io.airbyte.workers.internal.bookkeeping.events

import io.airbyte.featureflag.Connection
import io.airbyte.featureflag.Multi
import io.airbyte.featureflag.UseStreamStatusTracker2024
import io.airbyte.featureflag.Workspace
import io.airbyte.protocol.models.AirbyteMessage
import io.airbyte.protocol.models.AirbyteTraceMessage
import io.airbyte.workers.general.CachingFeatureFlagClient
import io.airbyte.workers.internal.bookkeeping.OldStreamStatusTracker
import io.micronaut.context.event.ApplicationEventListener
import jakarta.inject.Singleton

/**
 * Custom application listener that handles Airbyte Protocol [AirbyteMessage.Type.TRACE] messages of type
 * [AirbyteTraceMessage.Type.STREAM_STATUS] to track the status (or progress) of a stream
 * within replication. The listener handles the messages asynchronously to avoid blocking
 * replication.
 */
@Singleton
class AirbyteStreamStatusMessageEventListener(
  private val streamStatusTracker: OldStreamStatusTracker,
  private val ffClient: CachingFeatureFlagClient,
) :
  ApplicationEventListener<ReplicationAirbyteMessageEvent> {
  override fun onApplicationEvent(event: ReplicationAirbyteMessageEvent): Unit = streamStatusTracker.track(event)

  override fun supports(event: ReplicationAirbyteMessageEvent): Boolean {
    val ffCtx = Multi(listOf(Workspace(event.replicationContext.workspaceId), Connection(event.replicationContext.connectionId)))

    if (ffClient.boolVariation(UseStreamStatusTracker2024, ffCtx)) {
      return false
    }

    return with(event.airbyteMessage) {
      type == AirbyteMessage.Type.TRACE && trace.type == AirbyteTraceMessage.Type.STREAM_STATUS
    }
  }
}
